Sources/Datadog/Tracing/TracingFeature.swift

```diff
@@ -14,13 +19,156 @@ internal final class TracingFeature {
+                dateProvider: dateProvider
+            )
+
+            // NOTE: any change to tracing data format requires updating the tracing directory url to be unique
+            let dataFormat = DataFormat(
+                prefix: "",
+                suffix: "",
+                separator: "\n"
+            )
+
+            self.writer = FileWriter(dataFormat: dataFormat, orchestrator: orchestrator, queue: readWriteQueue)
+            self.reader = FileReader(dataFormat: dataFormat, orchestrator: orchestrator, queue: readWriteQueue)
+        }
+    }
+
+    /// Encapsulates upload stack setup for `TracingFeature`.
+    class Upload {
+        /// Uploads spans to server.
+        let uploader: DataUploadWorker
+
+        init(
+            storage: Storage,
+            appContext: AppContext,
+            performance: PerformancePreset,
+            httpClient: HTTPClient,
+            tracesUploadURLProvider: UploadURLProvider,
+            networkConnectionInfoProvider: NetworkConnectionInfoProviderType,
+            uploadQueue: DispatchQueue
+        ) {
+            let httpHeaders: HTTPHeaders
+            let uploadConditions: DataUploadConditions
+
+            if let mobileDevice = appContext.mobileDevice { // mobile device
+                httpHeaders = HTTPHeaders(
+                    headers: [
+                        .contentTypeHeader(contentType: .textPlainUTF8),
+                        .userAgentHeader(for: mobileDevice, appName: appContext.executableName, appVersion: appContext.bundleVersion)
+                    ]
+                )
+                uploadConditions = DataUploadConditions(
+                    batteryStatus: BatteryStatusProvider(mobileDevice: mobileDevice),
+                    networkConnectionInfo: networkConnectionInfoProvider
+                )
+            } else { // other device (i.e. iOS Simulator)
+                httpHeaders = HTTPHeaders(
+                    headers: [
+                        .contentTypeHeader(contentType: .textPlainUTF8)
+                        // UA http header will default to the one produced by the OS
+                    ]
+                )
+                uploadConditions = DataUploadConditions(
+                    batteryStatus: nil, // uploads do not depend on battery status
+                    networkConnectionInfo: networkConnectionInfoProvider
+                )
+            }
+
+            let dataUploader = DataUploader(
+                urlProvider: tracesUploadURLProvider,
+                httpClient: httpClient,
+                httpHeaders: httpHeaders
+            )
+
+            self.uploader = DataUploadWorker(
+                queue: uploadQueue,
+                fileReader: storage.reader,
+                dataUploader: dataUploader,
+                uploadConditions: uploadConditions,
+                delay: DataUploadDelay(performance: performance),
+                featureName: "tracing"
+            )
+        }
+    }
```

Comment:
could `Upload` class be turned into an `DataUploadWorker` extension?
does it do something extra than `DataUploadWorker`?

Reply:
The `Upload` struct segregates dependencies on features, configures `DataUploadWorker` with feature-related HTTP headers, creates data upload conditions based on device kind and names the feature:
```swift
if let mobileDevice = appContext.mobileDevice { // mobile device
    httpHeaders = HTTPHeaders(
        headers: [
            .contentTypeHeader(contentType: .textPlainUTF8),
            .userAgentHeader(for: mobileDevice, appName: appContext.executableName, appVersion: appContext.bundleVersion)
        ]
    )
    uploadConditions = DataUploadConditions(
        batteryStatus: BatteryStatusProvider(mobileDevice: mobileDevice),
        networkConnectionInfo: networkConnectionInfoProvider
    )
} else { // other device (i.e. iOS Simulator)
    httpHeaders = HTTPHeaders(
        headers: [
            .contentTypeHeader(contentType: .textPlainUTF8)
            // UA http header will default to the one produced by the OS
        ]
    )
    uploadConditions = DataUploadConditions(
        batteryStatus: nil, // uploads do not depend on battery status
        networkConnectionInfo: networkConnectionInfoProvider
    )
}

self.uploader = DataUploadWorker(
    // ...
    dataUploader: dataUploader,
    uploadConditions: uploadConditions,
    // ...
    featureName: "tracing"
)
```

I don't see a good reason to make `DataUploadWorker` to depend on more things than it should and perform this configuration inside it. Also, this will just move the code around and make mocking more difficult - now we just mock all the dependencies for the feature and it results with consistent `Upload` + `Storage` stacks, but if we make the feature depend on `DataUploadWorker` directly, we'd have to consider corner cases like "mocking Tracing feature with Logging upload stack".